### PR TITLE
Added find_stack_level to util.warnings

### LINKED ISSUE #12646: [FEATURE]  Add a find_stack_level to deprecation.py
### Problem description

The stacklevel of the warn in bokeh/util/deprecation.py is set to 2, making it hard to find the actual place when running the panel test suite. e.g.

``` python
../bokeh/src/bokeh/util/deprecation.py:78
  BokehDeprecationWarning: Positive was deprecated in Bokeh 3.0.0 and will be removed, use Positive(Int) instead.
```

### Feature description

My suggestion will be to copy `pandas.util._exceptions.find_stack_level`; which will give the following output.

``` python
panel/models/vtk.py:28
  BokehDeprecationWarning: Positive was deprecated in Bokeh 3.0.0 and will be removed, use Positive(Int) instead.
```

The function `find_stack_level` I suggest will look something like this:
``` python
def find_stack_level() -> int:
    """
    Find the first place in the stack that is not inside Bokeh.

    Inspired by: pandas.util._exceptions.find_stack_level
    """

    import bokeh

    pkg_dir = os.path.dirname(bokeh.__file__)

    # https://stackoverflow.com/questions/17407119/python-inspect-stack-is-slow
    frame = inspect.currentframe()
    n = 0
    while frame:
        fname = inspect.getfile(frame)
        if fname.startswith(pkg_dir):
            frame = frame.f_back
            n += 1
        else:
            break
    return n
``` 

### Potential alternatives

Leave it as it is. 

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by hoxbro**
All pull requests must have an associated issue in the issue tracker. If there
isn't one, please go open an issue describing the defect, deficiency or desired
feature. You can read more about our issue and PR processes in the
[wiki](https://github.com/bokeh/bokeh/wiki/BEP-1:-Issues-and-PRs-management).

- [x] issues: fixes #12646
- [ ] tests added / passed
- [ ] release document entry (if new feature or API change)

### CODE REVIEW

File: src/bokeh/util/deprecation.py [hunk 1]
```diff
@@ -18,6 +18,8 @@
 #-----------------------------------------------------------------------------
 
 # Standard library imports
+import os
+import inspect
```
Comment (hoxbro) [inline src/bokeh/util/deprecation.py]:
```suggestion
import inspect
import os
```

**Comment by bryevdv** [comment]
Hi @Hoxbro thanks for the PR! I do think this ought to go in a new module (e.g. `bokeh.util.exceptions`) since there are other things besides just deprecations that could benefit from this, e.g. 

https://github.com/bokeh/bokeh/blob/8fa898b959bbb5139985ea261ef4baf4402dd752/src/bokeh/core/has_props.py#L218

as well as several other places that use `warnings.warn` 

Alternatively, we could move the `warn` function and `find_stacklevel` to a new `bokeh.util.warning` module, and then use our custom `warn` function everywhere we currently use `warnings.warn`. Let me know what you think!

**Comment by hoxbro** [comment]
I don't have a strong opinion about which of the opinions to do. Can do both and will update the rest of the code accordingly. 

Though I will probably use  [`bokeh.util.warnings`](https://github.com/bokeh/bokeh/blob/branch-3.1/src/bokeh/util/warnings.py) instead of `bokeh.util.warning`. If that option is chosen.

File: src/bokeh/util/deprecation.py [hunk 2]
```diff
@@ -77,6 +82,29 @@ def deprecated(since_or_msg: Version | str,
 
     warn(message)
 
+def find_stack_level() -> int:
+    """
+    Find the first place in the stack that is not inside Bokeh.
```
Comment (bryevdv) [inline src/bokeh/util/deprecation.py]: nit: all the docstings in the repo start like this:
```suggestion
    """ Find the first place in the stack that is not inside Bokeh.
```

**Comment by bryevdv** [comment]
@Hoxbro that's fantastic, appreciate your help. Let's move the custom `warn` and the new `find_stack_level` to a `bokeh.util.warnings` module. There a quite a number of places that currently call `warnings.warn`. It's really up to you as to how many of these you want to tackle, but let's make a follow-up issue in case it's not all of them :) 
```
bokeh/io/export.py
328:            warnings.warn("Export method called with width or height argument on a non-Plot model. The size values will be ignored.")

bokeh/io/notebook.py
312:        warn("No document to push")
319:        warn("Cannot find a last shown plot to update. Call output_notebook() and show(..., notebook_handle=True) before push_notebook()")

bokeh/io/saving.py
142:        warn("save() called but no resources were supplied and output_file(...) was never called, defaulting to resources.CDN")
154:        warn("save() called but no title was supplied and output_file(...) was never called, using default title 'Bokeh Plot'")

bokeh/settings.py
812:    warnings.warn("BOKEH_SECRET_KEY is recommended to have at least 32 bytes of entropy chosen with a cryptographically-random algorithm")
817:    warnings.warn("BOKEH_SECRET_KEY must be set if BOKEH_SIGN_SESSIONS is set to True")

bokeh/command/util.py
130:                warnings.warn(DIRSTYLE_MAIN_WARNING)

bokeh/plotting/_tools.py
156:        warnings.warn(f"{','.join(repeated_tools)} are being repeated")

bokeh/plotting/graph.py
83:            warn("Converting node attributes labeled 'index' are skipped. "
104:            warn("Converting edge attributes labeled 'start' or 'end' are skipped. "
122:                warn("Node keys in 'layout_function' don't match node keys in the graph. "

bokeh/models/sources.py
211:                 lambda obj, name, data: warnings.warn(
416:            warnings.warn("Unable to find column '%s' in data source" % name)

bokeh/util/deprecation.py
45:def warn(message: str, stacklevel: int = 2) -> None:
46:    warnings.warn(message, BokehDeprecationWarning, stacklevel=stacklevel)
78:    warn(message)

bokeh/util/terminal.py
96:def warn(msg: str | None = None, label: str = "WARN") -> None:

bokeh/util/token.py
260:        warnings.warn('A secure pseudo-random number generator is not available '
263:            warnings.warn('A secure pseudo-random number generator is not available '

bokeh/models/plots.py
954:            warnings.warn(_LEGEND_EMPTY_WARNING % attr)

bokeh/embed/bundle.py
174:            warn('No Bokeh CSS Resources provided to template. If required you will need to provide them manually.')
177:            warn('No Bokeh JS Resources provided to template. If required you will need to provide them manually.')

bokeh/models/formatters.py
399:        warnings.warn(f"DatetimeFormatter scales now only accept a single format. Using the first prodvided: {fmt[0]!r} ")

bokeh/core/has_props.py
215:            warn(f"Properties {redeclared!r} in class {cls.__name__} were previously declared on a parent "
223:            warn(f"Overrides of {unused_overrides} in class {cls.__name__} does not override anything.", RuntimeWarn
```

File: src/bokeh/util/warnings.py [hunk 3]
```diff
@@ -23,13 +23,25 @@
 # Imports
 #-----------------------------------------------------------------------------
 
+# Standard library imports
+import inspect
+import os
+import warnings  # lgtm [py/import-and-import-from]
```
Comment (mattpap) [inline src/bokeh/util/warnings.py:29]: Probably this `lgtm` comment isn't necessary anymore.
```suggestion
import warnings
```

Comment (bryevdv) [reply src/bokeh/util/warnings.py:29]: definitely, but I am going to remove all of them in a PR later this week, so there's no need to task @Hoxbro more

**Comment by bryevdv** [comment]
Thanks for the comprehensive PR @Hoxbro !

**Review by bryevdv — APPROVED** [review]

**Pull request merged**